pyro/infer/abstract_infer.py

```diff
@@ -137,12 +67,43 @@ def _traces(self, *args, **kwargs):
         """
         raise NotImplementedError("inference algorithm must implement _traces")
 
+    def _marginal_single_site(self, site="_RETURN"):
+        sites_key = frozenset([site])
+        if sites_key in self._marginals:
+            return self._marginals[sites_key]
+        marginal = Empirical()
+        for tr, log_weight in zip(self.exec_traces, self.log_weights):
+            marginal.add(tr.nodes[site]["value"], log_weight=log_weight)
+        self._marginals[sites_key] = marginal
+        return marginal
+
+    def marginal(self, sites=None):
+        """
+        Return the marginal distribution for one or more sample sites from the
+        collection of execution traces. If multiple sites are specified, they
+        must have the same tensor shape. The result is cached for any future
+        calls.
+
+        :param sites: sample site whose marginal to return.
+        :return: ``dist.Empirical``
+        """
+        if sites is None:
+            sites = "_RETURN"
+        assert isinstance(sites, (list, str))
+        sites_key = tuple(sites) if isinstance(sites, list) else sites
+        if sites_key in self._marginals:
+            return self._marginals[sites_key]
+        marginal = Empirical()
+        for tr, log_weight in zip(self.exec_traces, self.log_weights):
+            value = tr.nodes[sites]["value"] if isinstance(sites, str) else \
+                torch.stack([tr.nodes[site]["value"] for site in sites], 0)
+            marginal.add(value, log_weight=log_weight)
+        self._marginals[sites_key] = marginal
+        return marginal
+
     def __call__(self, *args, **kwargs):
-        traces, logits = [], []
+        self._init()
         for tr, logit in poutine.block(self._traces)(*args, **kwargs):
-            traces.append(tr)
-            logits.append(logit)
-        logits = torch.stack(logits).squeeze()
-        logits -= log_sum_exp(logits)
-        ix = dist.Categorical(logits=logits).sample()
-        return traces[ix]
+            self.exec_traces.append(tr)
+            self.log_weights.append(logit)
+        return self
```

Comment:
Per design doc/discussion, should this return a sample trace with only latent variables?  That way you could use it for prediction as well via `replay`.  Thoughts?

Reply:
This will populate `exec_traces`, which can be accessed directly as `mcmc(data).exec_traces`. Returning `self` has the advantage that we can use it to marginalize too - like `mcmc(data).marginal("site")`. I am fine with returning exec_traces directly from this method itself, and don't have a strong opinion on this otherwise.